### PR TITLE
Read data-versions file to get the correct data file

### PR DESCRIPTION
**Pull request by pawbu**
As discussed we will now store information about data files with respect to each version in data_versions__x_x_x.json files. Example:

```
datastore/data_versions__1_0_0.json:
{
    "BOSTED": "BOSTED/BOSTED__1_0.parquet",
    "UTDANNING": "UTDANNING/UTDANNING__1_0",
    "ANNEN_VARIABEL": "ANNEN_VARIABEL/ANNEN_VARIABEL__1_0.parquet"
}
```

Data-service needs to find correct file and the specified dataset in that file.

### CODE REVIEW

File: data_service/adapters/storage/local.py [hunk 1]
```diff
@@ -14,13 +14,13 @@ def __init__(self, settings: config.LocalFileSettings = Depends(get_settings)):
...
-        return self.__create_download_path(dataset_name, dataset_version)
+    def get_file_path(self, path: str) -> str:
+        return f"{self.settings.DATASTORE_DIR}/data/{path}"
 
-    def __create_download_path(self, dataset_name: str, version: str) -> str:
-        path = (
-            f"{self.settings.DATASTORE_DIR}/data/"
-            f"{dataset_name}/{dataset_name}__{version}"
+    def get_data_versions(self, version: str) -> str:
+        version_underscored = version.replace('.', '_')[:3]
```
Comment (DanielElisenberg) [inline data_service/adapters/storage/local.py:21]: We have to remember to not bother making a new "data_versions"-file on bump if the bump is only a PATCH

**Comment by vaskalan** [comment]
What about this?
```
datastore/data_versions__1_0_0.json:
{
    "BOSTED": "BOSTED__1_0.parquet",
    "UTDANNING": "UTDANNING__1_0",
    "ANNEN_VARIABEL": "ANNEN_VARIABEL__1_0.parquet"
}
```
Suggestion : Remove directory as it will always bed the same as variable name

**Review by vaskalan — APPROVED** [review]

**Review by DanielElisenberg — COMMENTED** [review]
I wonder if maybe it would be better if the data_versions-file is abstracted away from the processor module completely? 
You just ask for a file path from the file adapter with a dataset_name + version and it will handle everything with the file etc. and return the complete path/appropriate exceptions

File: data_service/adapters/storage/local.py (hunk 1 above)
Comment (vaskalan) [reply data_service/adapters/storage/local.py:21]: If last datastore version is i.e. 1.2.0 and you bump a PATCH then the new datastore version will be 1.2.1.
Won't we make a data_versions__1_2_1 then?

Comment (pawbu) [reply data_service/adapters/storage/local.py:21]: Yes, I agree with @vaskalan. We need to create it for every version that the client can query.

**Comment by pawbu** [comment]
> I wonder if maybe it would be better if the data_versions-file is abstracted away from the processor module completely? You just ask for a file path from the file adapter with a dataset_name + version and it will handle everything with the file etc. and return the complete path/appropriate exceptions

Revisiting this idea now. Seems a bit better, yes.

File: data_service/adapters/storage/local.py (hunk 1 above)
Comment (vaskalan) [reply data_service/adapters/storage/local.py:21]: The naming standard should be `data_versions__1_0.json` if we decide not to crate a new data_versions feil for PATCH.

Comment (pawbu) [reply data_service/adapters/storage/local.py:21]: Yes, most probably there will not be a case for PATCH with new/changed/revised data. With this change the bumping will need to check bump type and create these data-versions files only if type is ADD or CHANGE_DATA. There will be fewer `data_versions__x_x.json` files though.

Comment (pawbu) [reply data_service/adapters/storage/local.py:21]: or bump type is MAJOR or MINOR

Comment (DanielElisenberg) [reply data_service/adapters/storage/local.py:21]: Yes, I agree its much cleaner to generate one for each datafile. I initially thought we were not going for this based on the data files themselves having just __MAJOR_MINOR in their file name, and the line of code this thread is connected to.
Generating one data_version__M_m_p.json for each bump regardless of bump-type seems like the cleanest/simplest solution though, so I absolutely agree.

File: data_service/exceptions.py [hunk 2]
```diff
@@ -0,0 +1,2 @@
+class NotFoundException(Exception):
```
Comment (DanielElisenberg) [inline data_service/exceptions.py:1]: :100:

**Review by DanielElisenberg — APPROVED** [review]
Looks great 👍🏻 Only thing missing is the discussion on line 20 in. data_service/adapters/storage/local.py
where
`version_underscored = version.replace('.', '_')[:3] ` ex.: 1_2
should be rewritten to:
`version_underscored = version.replace('.', '_')[:5]` ex.:  1_2_3

**Comment by pawbu** [comment]
> Looks great 👍🏻 Only thing missing is the discussion on line 20 in. data_service/adapters/storage/local.py where `version_underscored = version.replace('.', '_')[:3] ` ex.: 1_2 should be rewritten to: `version_underscored = version.replace('.', '_')[:5]` ex.: 1_2_3

You're absolutely right. I totally missed this one (+the data-versions file name as well).

**Comment by pawbu** [comment]
There is one more case that we need: 0.0.0 version. See my last commit https://github.com/statisticsnorway/microdata-data-service/pull/50/commits/8e4801404597af42a3081282259aaad33fc0b637

File: data_service/adapters/storage/local.py [hunk 3]
```diff
@@ -14,13 +16,36 @@ def __init__(self, settings: config.LocalFileSettings = Depends(get_settings)):
         self.log = logging.getLogger(__name__ + '.LocalFileAdapter')
         self.settings = settings
 
-    def get_file(self, dataset_name: str, version: str) -> str:
-        dataset_version = version.replace('.', '_')[:3]
-        return self.__create_download_path(dataset_name, dataset_version)
+    def get_file_path(self, data_structure_name: str, version: str) -> str:
```
[the remark this answers is not in the document]
Comment (pawbu) [reply data_service/adapters/storage/local.py:19]: Let's be technology-agnostic, `get_data_file_path`?

**Review by vaskalan — APPROVED** [review]

File: data_service/adapters/storage/local.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (DanielElisenberg) [reply data_service/adapters/storage/local.py:19]: data_file_path and parquet_file_path are both equally good in my opinion 👍🏻 Can't go wrong

File: data_service/adapters/storage/local.py [hunk 4]
```diff
@@ -20,7 +20,7 @@ def get_parquet_file_path(self, data_structure_name: str, version: str) -> str:
         version_underscored = version.replace('.', '_')[:5]
         path_prefix = f"{self.settings.DATASTORE_DIR}/data/{data_structure_name}"
 
-        if version == "0.0.0":
+        if version.startswith("0.0.0") or version == "draft":
             parquet_file_path = f"{data_structure_name}__{version_underscored[:3]}"
```
Comment (DanielElisenberg) [inline data_service/adapters/storage/local.py:24]: This will still look for data_version files with only 2 version values ☝🏻

Comment (DanielElisenberg) [reply data_service/adapters/storage/local.py:24]: never mind this is the parquet file path 🤦🏻 my bad

**Review by DanielElisenberg — APPROVED** [review]
LGTM :100:

File: data_service/adapters/storage/local.py (hunk 4 above)
Comment (pawbu) [reply data_service/adapters/storage/local.py:24]: It was actually a good point, but it will fail on line 20 🤕

Comment (DanielElisenberg) [reply data_service/adapters/storage/local.py:24]: Ah, true! An unintentional good point 😅 I'll take the win 🏅

**Pull request merged**